web-api/src/storage/postgres/client.rs

```diff
@@ -0,0 +1,329 @@
+    ///
+    /// Passwords in the URL will be ignored, do not set the
+    /// db password with the db url.
+    base_url: String,
+
+    /// Override port from base url.
+    port: Option<u16>,
+
+    /// Override user from base url.
+    user: Option<String>,
+
+    /// Sets the password.
+    #[serde(serialize_with = "serialize_redacted")]
+    password: Secret<String>,
+
+    /// Override db from base url.
+    db: Option<String>,
+
+    /// Override default application name from base url.
+    application_name: Option<String>,
+
+    /// If true skips running db migrations on start up.
+    skip_migrations: bool,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            base_url: "postgres://user:pw@localhost:5432/xayn".into(),
+            port: None,
+            user: None,
+            password: String::from("pw").into(),
+            db: None,
+            application_name: option_env!("CARGO_BIN_NAME").map(|name| format!("xayn-web-{name}")),
+            skip_migrations: false,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct DatabaseBuilder {
+    pool: Pool<Postgres>,
+}
+
+impl DatabaseBuilder {
+    pub(crate) async fn close(&self) {
+        self.pool.close().await;
+    }
+
+    pub(crate) fn build_for(&self, tenant_id: TenantId) -> Result<Database, Error> {
+        Ok(Database {
+            pool: self.pool.clone(),
+            tenant_db_name: db_name_for_tenant_id(tenant_id).map_err(InternalError::from_std)?,
+        })
+    }
+}
+
+fn db_name_for_tenant_id(tenant_id: TenantId) -> Result<QuotedIdentifier, InvalidQuotedIdentifier> {
+    format!("t:{tenant_id}").try_into()
+}
+
+#[derive(Debug)]
+pub(crate) struct Database {
+    pool: Pool<Postgres>,
+    tenant_db_name: QuotedIdentifier,
+}
+
+impl Database {
+    #[instrument]
+    pub(crate) async fn build(
+        config: &Config,
+        enable_legacy_tenant: bool,
+    ) -> Result<DatabaseBuilder, SetupError> {
+        let options = Self::build_connection_options(config)?;
+        info!("starting postgres setup");
+        let pool = PoolOptions::new()
+            .after_release(|conn, _metadata| {
+                async {
+                    sqlx::query("RESET ROLE;").execute(conn).await?;
+                    Ok(true)
+                }
+                .boxed()
+            })
+            .connect_with(options)
+            .await?;
+
+        if !config.skip_migrations {
+            sqlx::migrate!().run(&pool).await?;
+
+            //FIXME: This will be handled nicer once schema migration has it's own
+            //       API the follow up PR
+            if enable_legacy_tenant {
+                Self::legacy_tenant_setup(&pool).await?;
+            }
+        }
+
+        Ok(DatabaseBuilder { pool })
+    }
+
+    #[instrument(err)]
+    async fn legacy_tenant_setup(pool: &Pool<Postgres>) -> Result<(), SetupError> {
+        //FIXME: THIS CODE WILL BE COMPLETELY REPLACED IN FOLLOW UP PR
```

Comment:
:wrench: 
This code seems quite problematic, it mentions hacks, workaround and known problems during testing.
I don't think we need any of this.
We should not touch the single-tenant deployment if possible. In this case, is much simpler to avoid sending `set role` when in single mode than trying to do any database migration. None of the current systems will be migrated any time soon to the multi-tenant mode.

Reply:
> I don't think we need any of this.

This code only exists because this set of deeply interconnected changes is split into different PRs and each must work by itself as far as possible.

So if we merge the various PRs in this story into each other and then merge it into master we indeed don't need this.

> In this case, is much simpler to avoid sending set role

Long term (in the follow up PRs) no, not at all.

> None of the current systems will be migrated any time soon to the multi-tenant mode.

This isn't about that at all, this is about making the code in the follow up PRs simpler and easier
to maintain. 

> problematic

This was problematic when I had been using the same id for all legacy tenants, by now it isn't problematic anymore as there won't be any collisions/race conditions as all tenants including legacy tenants have unique ids across all databases (a thing we need anyway).

Through I had considered changing the code to no use users. I.e. instead of using `SET ROLE` it would use `SET search_path TO {tenant_schema_name};` instead.

But this would remove some security protections against query injection attacks accessing data from other users and would make future migrations much harder so I'm not sure about this.